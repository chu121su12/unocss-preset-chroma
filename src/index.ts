import type { Preset } from 'unocss'
import { escapeRegExp } from 'unocss'
import { parseColor } from '@unocss/preset-mini'
import chroma from 'chroma-js'

/**
 * @public
 */
export interface ChromaOptions {
  /**
   * Class prefix for matching gradient rules.
   *
   * @defaultValue `chroma-`
   */
  prefix?: string
  /**
   * Number of gradient steps to generate.
   *
   * @defaultValue 7
   */
  steps?: number
}

const colorsReStr = /(?:\w+(?:-\d+)?|(?:#|hex-)[0-9a-f]{3,8})(?:\/\d+)?/.source
const colorBodyRe = new RegExp(`^-(${colorsReStr})`)

export function presetChroma(options: ChromaOptions = {}): Preset {
  const prefix = escapeRegExp(options.prefix ?? 'chroma-')
  const defaultSteps = options.steps ?? 7
  const defaultShapes = {
    linear: '0deg',
    radial: 'circle at 50% 50%',
    conic: 'at 50% 50%',
  }

  return {
    name: 'unocss-preset-chroma',
    rules: [
      [
        new RegExp(`^${prefix}(?:((?:linear|radial|conic))-)?(?:(\\d+)-)?(rgb|lab|hsl|lch)((?:-${colorsReStr}){2,})$`),
        ([, type = 'linear', precision, mode, body], { theme }) => {
          let steps = parseInt(precision || '0', 10)
          if (Number.isNaN(steps) || steps < 1)
            steps = defaultSteps
          const stops = resolveStops(steps, mode, body, theme)
          if (stops != null) {
            return {
              '--un-gradient-shape': defaultShapes[type],
              '--un-gradient-stops': stops.join(','),
              '--un-gradient': 'var(--un-gradient-shape), var(--un-gradient-stops)',
              'background-image': `${type}-gradient(var(--un-gradient))`,
            }
          }
        },
      ],

      [
        new RegExp(`^${prefix}stops-(?:(\\d+)-)?(rgb|lab|hsl|lch)((?:-${colorsReStr}){2,})$`),
        ([, precision, mode, body], { theme }) => {
          let steps = parseInt(precision || '0', 10)
          if (Number.isNaN(steps) || steps < 1)
            steps = defaultSteps
          const stops = resolveStops(steps, mode, body, theme)
          if (stops != null)
            return { '--un-gradient-stops': stops.join(',') }
        },
      ],

      [
        new RegExp(`^${prefix}shape-\\[(.+)\\]$`),
        ([, s]) => ({
          '--un-gradient-shape': s.replace(/_/g, ' '),
          '--un-gradient': `var(--un-gradient-shape), var(--un-gradient-stops)`,
        }),
      ],

      [
        new RegExp(`^${prefix}((?:repeating-)?(?:linear|radial|conic))$`),
        ([, s]) => ({
          'background-image': `${s}-gradient(var(--un-gradient, var(--un-gradient-stops, rgba(255, 255, 255, 0))))`,
        }),
      ],
    ],
  }
}

export default presetChroma

function resolveStops(steps: number, mode: string, body: string, theme: any) {
  const cs: any[] = []
  while (true) {
    const m = body.match(colorBodyRe)
    if (!m)
      break

    const c = parseColor(m[1] ?? '', theme)
    if (!(c && c.color))
      return

    cs.push(c)
    body = body.replace(colorBodyRe, '')
  }

  if (body.length)
    return

  const g: string[][] = []
  const s = cs.length - 1
  for (let i = 0; i < s; i++) {
    const last = i === s - 1
    const r = resolveChroma(last ? steps : steps + 1, mode, cs[i].color, cs[i + 1].color)
      .map(([c, p])  => [parseColor(c, theme)?.cssColor?.components?.join(','), Math.floor(i * 100 / s + p / s)])
      .map(([c, p]) => `rgb(${c}) ${p}%`)
    g.push(last ? r : r.slice(0, -1))
  }

  return g
}

function resolveChroma(steps: number, mode: string, start: string, end: string) {
  const scale = chroma.scale([start, end]).mode(mode).correctLightness()
  return [...Array(steps).keys()].map((_, i) => {
    const t = i / (steps - 1)
    return [scale(t).hex(), t * 100]
  })
}
