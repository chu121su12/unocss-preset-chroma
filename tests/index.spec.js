import { describe, expect, it } from 'vitest'
import { createGenerator, presetUno } from 'unocss'
import { presetChroma } from '../src/index'

describe('chroma preset', async () => {
  it('generate gradient css', async () => {
    const generator = createGenerator({
      presets: [
        presetUno(),
        presetChroma(),
      ],
    })

    const target = [
      'chroma-rgb-#ff0-#00f chroma-shape-[90deg]',
      'chroma-conic-hsl-#a08df7-#fff-#ffe261',
      'chroma-linear-lab-#000080-#ffff00',
      'chroma-linear-hsl-#008000-#ff00ff',
      'chroma-linear-lch-#800000-#00ffff',
      'chroma-linear-direction-30',
    ].join(' ')

    const { css } = await generator.generate(target, { preflights: false })
    expect(css).toMatchSnapshot()
  })
})
