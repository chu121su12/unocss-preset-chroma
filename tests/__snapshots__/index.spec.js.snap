// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`chroma preset > generate gradient css 1`] = `
"/* layer: default */
.chroma-conic-hsl-\\\\#a08df7-\\\\#fff-\\\\#ffe261{--un-gradient-shape:at 50% 50%;--un-gradient-stops:rgb(160,141,247) 0%,rgb(174,158,248) 7%,rgb(187,173,249) 14%,rgb(201,190,250) 21%,rgb(214,206,252) 28%,rgb(228,223,253) 35%,rgb(241,238,254) 42%,rgb(255,255,255) 50%,rgb(255,251,230) 58%,rgb(255,246,208) 66%,rgb(255,241,181) 75%,rgb(255,237,154) 83%,rgb(255,232,127) 91%,rgb(255,226,97) 100%;--un-gradient:var(--un-gradient-shape), var(--un-gradient-stops);background-image:conic-gradient(var(--un-gradient));}
.chroma-linear-hsl-\\\\#008000-\\\\#ff00ff{--un-gradient-shape:0deg;--un-gradient-stops:rgb(0,128,0) 0%,rgb(0,135,21) 16%,rgb(0,140,41) 33%,rgb(216,0,249) 50%,rgb(228,0,251) 66%,rgb(241,0,253) 83%,rgb(255,0,255) 100%;--un-gradient:var(--un-gradient-shape), var(--un-gradient-stops);background-image:linear-gradient(var(--un-gradient));}
.chroma-linear-lab-\\\\#000080-\\\\#ffff00{--un-gradient-shape:0deg;--un-gradient-stops:rgb(0,0,128) 0%,rgb(86,43,123) 16%,rgb(127,82,116) 33%,rgb(163,123,107) 50%,rgb(195,166,93) 66%,rgb(225,210,70) 83%,rgb(255,255,0) 100%;--un-gradient:var(--un-gradient-shape), var(--un-gradient-stops);background-image:linear-gradient(var(--un-gradient));}
.chroma-linear-lch-\\\\#800000-\\\\#00ffff{--un-gradient-shape:0deg;--un-gradient-stops:rgb(128,0,0) 0%,rgb(136,67,0) 16%,rgb(134,111,0) 33%,rgb(122,151,46) 50%,rgb(99,188,109) 66%,rgb(59,223,180) 83%,rgb(0,255,255) 100%;--un-gradient:var(--un-gradient-shape), var(--un-gradient-stops);background-image:linear-gradient(var(--un-gradient));}
.chroma-rgb-\\\\#ff0-\\\\#00f{--un-gradient-shape:0deg;--un-gradient-stops:rgb(255,255,0) 0%,rgb(223,223,32) 16%,rgb(192,192,63) 33%,rgb(161,161,94) 50%,rgb(129,129,126) 66%,rgb(95,95,160) 83%,rgb(0,0,255) 100%;--un-gradient:var(--un-gradient-shape), var(--un-gradient-stops);background-image:linear-gradient(var(--un-gradient));}
.chroma-shape-\\\\[90deg\\\\]{--un-gradient-shape:90deg;--un-gradient:var(--un-gradient-shape), var(--un-gradient-stops);}"
`;
